### PR TITLE
8274784: jshell: Garbled character was displayed by System.out.println(...) on Japanese Windows

### PR DESCRIPTION
**Pull request by takiguc**
JEP-400 was implemented by JDK18-b13.
After JDK18-b13, garbled character was displayed by following code on Japanese Windows' command prompt.

System.out.println("\u3042")

Japanese "A" should be display ed, but garbled character was displayed.
Also saved jshell command list did not work as expected if Japanese character was there.

Following issue has some information
8274544: Langtools command's usage were garbled on Japanese Windows #5771
https://github.com/openjdk/jdk/pull/5771
This issue also happens on Linux ja_JP.eucjp locale.
RemoteExecutionControl.java change is required for this issue.

Also we cannot input Japanese character on Linux ja_JP.eucjp locale terminal.
AbstractTerminal.java change is required for this issue.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274784](https://bugs.openjdk.java.net/browse/JDK-8274784): jshell: Garbled character was displayed by System.out.println(...) on Japanese Windows

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6505/head:pull/6505` \
`$ git checkout pull/6505`

Update a local copy of the PR: \
`$ git checkout pull/6505` \
`$ git pull https://git.openjdk.java.net/jdk pull/6505/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6505`

View PR using the GUI difftool: \
`$ git pr show -t 6505`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6505.diff">https://git.openjdk.java.net/jdk/pull/6505.diff</a>

</details>

### CODE REVIEW

File: src/jdk.internal.le/share/classes/jdk/internal/org/jline/terminal/impl/AbstractTerminal.java [hunk 1]
```diff
@@ -55,7 +56,8 @@ public AbstractTerminal(String name, String type) throws IOException {
     public AbstractTerminal(String name, String type, Charset encoding, SignalHandler signalHandler) throws IOException {
         this.name = name;
         this.type = type != null ? type : "ansi";
-        this.encoding = encoding != null ? encoding : Charset.defaultCharset();
+        this.encoding = encoding != null ? encoding :
+            Charset.forName(new OutputStreamWriter(System.out).getEncoding(), Charset.defaultCharset());
```
Comment (naotoj) [inline src/jdk.internal.le/share/classes/jdk/internal/org/jline/terminal/impl/AbstractTerminal.java:60]: I don't think `OutputStreamWriter` instance is needed here. 
```
System.out instanceof PrintStream ps ? ps.charset() : Charset.defaultEncoding();
```
would suffice.

File: src/jdk.jshell/share/classes/jdk/jshell/execution/RemoteExecutionControl.java [hunk 2]
```diff
@@ -63,13 +65,19 @@ public static void main(String[] args) throws Exception {
...
         input.put("in", System::setIn);
         forwardExecutionControlAndIO(new RemoteExecutionControl(), inStream, outStream, outputs, input);
     }
 
+    // Get PrintStream charset
+    private static Charset getPrintStreamCharset(PrintStream ps) {
+        return Charset.forName(new OutputStreamWriter(ps).getEncoding(),
+            Charset.defaultCharset());
+    }
+
```
Comment (naotoj) [inline src/jdk.jshell/share/classes/jdk/jshell/execution/RemoteExecutionControl.java:80]: I'd prefer not to introduce this static method, but just embed it into each location. The other comment in `AbstractTerminal.java` applies here too.

**Comment by takiguc** [comment]
Thanks @naotoj .
I appreciate your suggestion.
I misread PrintStream.charset() spec.

**Review by naotoj — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**